apps/heft/src/plugins/NodeServicePlugin.ts

```diff
@@ -0,0 +1,343 @@
+          bundle.hooks.run.tapPromise(PLUGIN_NAME, async () => {
+            await this._runCommandAsync(heftSession, heftConfiguration);
+          });
+        });
+
+        build.hooks.compile.tap(PLUGIN_NAME, (compile: ICompileSubstage) => {
+          compile.hooks.afterEachIteration.tap(PLUGIN_NAME, this._compileHooks_afterEachIteration);
+        });
+      }
+    });
+  }
+
+  private async _runCommandAsync(
+    heftSession: HeftSession,
+    heftConfiguration: HeftConfiguration
+  ): Promise<void> {
+    this._logger.terminal.writeLine(`Starting Node service...`);
+
+    this._restartChild();
+  }
+
+  private _compileHooks_afterEachIteration = (): void => {
+    try {
+      // We've recompiled, so try launching again
+      this._childProcessFailed = false;
+
+      if (this._state === State.Stopped) {
+        // If we are already stopped, then extend the timeout
+        this._scheduleRestart(this._configuration.waitBeforeRestartMs);
+      } else {
+        this._stopChild();
+      }
+    } catch (error) {
+      console.error('UNCAUGHT: ' + error.toString());
+    }
+  };
+
+  private _restartChild(): void {
+    if (this._state !== State.Stopped) {
+      throw new InternalError('Invalid state');
+    }
+
+    this._state = State.Running;
+    this._clearTimeout();
+
+    this._logger.terminal.writeLine('Invoking command: ' + JSON.stringify(this._shellCommand!));
+
+    this._activeChildProcess = child_process.spawn(this._shellCommand!, {
+      shell: true,
+      stdio: ['inherit', 'inherit', 'inherit'],
+      ...SubprocessTerminator.RECOMMENDED_OPTIONS
+    });
+    SubprocessTerminator.killProcessTreeOnExit(
+      this._activeChildProcess,
+      SubprocessTerminator.RECOMMENDED_OPTIONS
+    );
+
+    const childPid: number = this._activeChildProcess.pid;
+    this._logger.terminal.writeVerboseLine(`Started child ${childPid}`);
+
+    this._activeChildProcess.on('close', (code: number, signal: string): void => {
+      // The 'close' event is emitted after a process has ended and the stdio streams of a child process
+      // have been closed. This is distinct from the 'exit' event, since multiple processes might share the
+      // same stdio streams. The 'close' event will always emit after 'exit' was already emitted,
+      // or 'error' if the child failed to spawn.
+
+      if (this._state === State.Running) {
+        this._logger.terminal.writeWarningLine(
+          `Child #${childPid} terminated unexpectedly code=${code} signal=${signal}`
+        );
+        this._childProcessFailed = true;
+        this._transitionToStopped();
+        return;
+      }
+
+      if (this._state === State.Stopping || this._state === State.Killing) {
+        this._logger.terminal.writeVerboseLine(
+          `Child #${childPid} terminated successfully code=${code} signal=${signal}`
+        );
+        this._transitionToStopped();
+        return;
+      }
+    });
+
+    // This is event requires Node.js >= 15.x
```

Comment:
Will this get ignored on <15, or will it cause problems?

Reply:
Node.js silently ignores unknown signal names.  (Of course it does heheh.)

However this comment is incorrect actually.  I'm testing with Node 12 and the event does get fired.  I don't remember where I read about that, but the Node.js docs say `Added in: v0.1.90`.

Reply:
Okay I found it, it is actually the ["spawn"](https://nodejs.org/api/child_process.html#child_process_event_spawn) event that requires Node 15.